src/AppInstallerCommonCore/Settings.cpp

```diff
@@ -347,7 +350,132 @@ namespace AppInstaller::Settings
+
+                    bool tempSettingsTrusted = false;
+                    {
+                        // Extra scope to release the file lock right after trust validation.
+                        std::unique_ptr<Msix::WriteLockedMsixFile> tempFileLock;
+                        tempSettingsTrusted = LockSettingsFile(tempSettingsPath, tempFileLock);
+                    }
+
+                    if (!tempSettingsTrusted)
+                    {
+                        AICLI_LOG(Core, Error, << "Failed to open remote settings, downloaded temp remote settings not trusted.");
+                        THROW_HR(APPINSTALLER_CLI_ERROR_REMOTE_SETTINGS_NOT_TRUSTED);
+                    }
+
+                    // Move and lock the newly downloaded settings file.
+                    std::filesystem::rename(tempSettingsPath, m_settingsFile);
+                    if (!LockSettingsFile(m_settingsFile, m_fileLock))
+                    {
+                        AICLI_LOG(Core, Error, << "Failed to open remote settings, downloaded remote settings not trusted.");
+                        THROW_HR(APPINSTALLER_CLI_ERROR_REMOTE_SETTINGS_NOT_TRUSTED);
+                    }
+                }
+
+                AICLI_LOG(Repo, Info, << "Opened trusted remote settings.");
+            }
+
+            std::unique_ptr<std::istream> Get() override
+            {
+                if (std::filesystem::exists(m_settingsFile))
+                {
+                    auto result = std::make_unique<std::ifstream>(m_settingsFile, std::ios::binary);
+                    THROW_LAST_ERROR_IF(result->fail());
+                    return result;
+                }
+                else
+                {
+                    return {};
+                }
+            }
+
+            bool Set(std::string_view) override
+            {
+                THROW_HR(E_UNEXPECTED);
+            }
+
+            void Remove() override
+            {
+                THROW_HR(E_UNEXPECTED);
+            }
+
+            std::filesystem::path PathTo() override
+            {
+                THROW_HR(E_UNEXPECTED);
+            }
+
+        private:
+            // Try to use cached copy of the settings file if applicable.
+            // Returns true if the settings file is trusted, up to date and successfully locked. Otherwise false.
+            bool TryLockCachedSettingsFile()
+            {
+                if (!std::filesystem::exists(m_settingsFile))
+                {
+                    AICLI_LOG(Core, Info, << "Cached remote settings file does not exist.");
+                    return false;
+                }
+
+                if (!LockSettingsFile(m_settingsFile, m_fileLock))
+                {
+                    return false;
+                }
+
+                Msix::MsixInfo remoteInfo{ s_RemoteSettings_Url };
+                if (remoteInfo.IsNewerThan(m_settingsFile))

```

Comment:
One thing this does not account for is an attempt at a replay attack where we never try to move forward (the attacker makes us see the same old file forever).

To remain trusted, we would need to have some way to time out an older file.  Encoding the creation date in the version is already being done, and this could be used to ensure that the current (and remote) packages are up to date.

Additionally, for performance we probably want to not even check for a remote update unless there is a failure (or maybe based on time).